File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/schema/GQLProjectEntityDecorationListFieldContributor.java
package net.nemerosa.ontrack.graphql.schema;

import graphql.schema.DataFetcher;
import graphql.schema.GraphQLFieldDefinition;
import net.nemerosa.ontrack.graphql.support.GraphqlUtils;
import net.nemerosa.ontrack.model.structure.DecorationService;
import net.nemerosa.ontrack.model.structure.ProjectEntity;
import net.nemerosa.ontrack.model.structure.ProjectEntityType;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static graphql.Scalars.GraphQLString;
import static graphql.schema.GraphQLArgument.newArgument;
import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;

@Component
public class GQLProjectEntityDecorationListFieldContributor implements GQLProjectEntityFieldContributor {

    private final DecorationService decorationService;
    private final GQLTypeDecoration decoration;

    @Autowired
    public GQLProjectEntityDecorationListFieldContributor(DecorationService decorationService, GQLTypeDecoration decoration) {
        this.decorationService = decorationService;
        this.decoration = decoration;
    }

    @Override
    public List<GraphQLFieldDefinition> getFields(Class<? extends ProjectEntity> projectEntityClass, ProjectEntityType projectEntityType) {
        return Collections.singletonList(
                newFieldDefinition()
                        .name("decorations")
                        .description("List of decorations")
                        .argument(
                                newArgument()
                                        .name("type")
                                        .description("Fully qualified name of the decoration type")
                                        .type(GraphQLString)
                                        .build()
                        )
                        .type(GraphqlUtils.stdList(decoration.getTypeRef()))
                        .dataFetcher(projectEntityDecorationsDataFetcher(projectEntityClass))
                        .build()
        );
    }

    private DataFetcher projectEntityDecorationsDataFetcher(Class<? extends ProjectEntity> projectEntityClass) {
        return environment -> {
            Object o = environment.getSource();
            if (projectEntityClass.isInstance(o)) {
                // Filters
                Optional<String> typeFilter = GraphqlUtils.getStringArgument(environment, "type");
                // Gets the raw list
                return decorationService.getDecorations((ProjectEntity) o).stream()
                        // Filter by type
                        .filter(property -> typeFilter
                                .map(typeFilterName -> StringUtils.equals(
                                        typeFilterName,
                                        property.getDecorationType()
                                ))
                                .orElse(true)
                        )
                        // OK
                        .collect(Collectors.toList());
            } else {
                return null;
            }
        };
    }
}

File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/schema/GQLProjectEntityDecorationListFieldContributor.kt
package net.nemerosa.ontrack.graphql.schema

import graphql.Scalars
import graphql.schema.DataFetcher
import graphql.schema.GraphQLArgument
import graphql.schema.GraphQLFieldDefinition
import net.nemerosa.ontrack.graphql.support.listType
import net.nemerosa.ontrack.model.structure.Decoration
import net.nemerosa.ontrack.model.structure.DecorationService
import net.nemerosa.ontrack.model.structure.ProjectEntity
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import org.springframework.stereotype.Component

@Component
class GQLProjectEntityDecorationListFieldContributor(
        private val decorationService: DecorationService,
        private val decoration: GQLTypeDecoration,
) : GQLProjectEntityFieldContributor {

    override fun getFields(
            projectEntityClass: Class<out ProjectEntity>,
            projectEntityType: ProjectEntityType,
    ): List<GraphQLFieldDefinition>? {
        return listOf<GraphQLFieldDefinition>(
                GraphQLFieldDefinition.newFieldDefinition()
                        .name("decorations")
                        .description("List of decorations")
                        .argument(
                                GraphQLArgument.newArgument()
                                        .name("type")
                                        .description("Fully qualified name of the decoration type")
                                        .type(Scalars.GraphQLString)
                                        .build()
                        )
                        .type(listType(decoration.typeRef))
                        .dataFetcher(projectEntityDecorationsDataFetcher(projectEntityClass))
                        .build()
        )
    }

    private fun projectEntityDecorationsDataFetcher(projectEntityClass: Class<out ProjectEntity>) =
            DataFetcher { environment ->
                val o = environment.getSource<Any>()
                if (projectEntityClass.isInstance(o)) {
                    // Filters
                    val typeFilter: String? = environment.getArgument("type")
                    // Gets the raw list
                    decorationService.getDecorations(o as ProjectEntity)
                            .filter { property: Decoration<*> ->
                                typeFilter?.let {
                                    it == property.decorationType
                                } ?: true
                            }
                } else {
                    return@DataFetcher null
                }
            }
}
